usr/man/_vax.d/man3/ctime.3
.\" SCCSID: @(#)ctime.3	8.2	10/31/90
.TH ctime 3
.SH Name
ctime, localtime, gmtime, asctime, difftime, mktime, timezone, tzset \- date and time functions 
.SH Syntax
As shown, the 
.PN ctime ,
.PN localtime ,
.PN gmtime ,
.PN asctime ,
.PN difftime, 
.PN mktime ,
and
.PN tzset
calls are common to both the non-System V environment and the
System V environment.
.SS Common to Both Environments
.nf
.B #include <time.h>
.PP
.B void tzset()
.PP
.B char *ctime(\fIclock\fP)
.B time_t *\fIclock\fP;
.PP
.B char *asctime(\fItm\fP)
.B struct \fItm\fP *\fItm\fP;
.PP
.B struct \fItm\fP *localtime(\fIclock\fP)
.B time_t *\fIclock\fP;
.PP
.B struct \fItm\fP *gmtime(\fIclock\fP)
.B time_t *\fIclock\fP;
.PP
.B double difftime(\fItime1\fP, \fItime0\fP)
.B time_t \fItime1\fP, \fItime0\fP;
.PP
.B time_t mktime(\fItimeptr\fP)
.B struct \fItm\fP *\fItimeptr\fP;
.PP
.B extern char *tzname[2];
.SS BSD Environment Only
.B char *timezone(\fIzone\fP, \fIdst\fP)
.SS System V and POSIX Environments Only
.B extern long timezone;
.PP
.B extern int daylight;
.fi
.SH Description
.NXR "ctime subroutine (standard C)"
.NXR "localtime subroutine (standard C)"
.NXR "gmtime subroutine (standard C)"
.NXR "asctime subroutine (standard C)"
.NXR "timezone subroutine"
.NXR "tzet subroutine"
.NXR "time" "converting to ASCII"
.NXR "date" "converting to ASCII"
.PP
The 
.PN tzset
call uses the value of the environment variable TZ to
set up the time conversion information used by 
.PN localtime .
.PP
If TZ does not appear in the environment, the file
.PN /etc/zoneinfo/localtime 
is used by 
.PN localtime .
If
this file fails for any reason, the Greenwich Mean Time 
(GMT) offset as provided by the
kernel is used.  In this case, Daylight Savings Time
(DST) is ignored, resulting in the time
being incorrect by some amount if DST is currently in effect.  If
this fails for any reason, GMT is used.
.PP
If TZ appears in the environment but its value is a null string,
GMT is used; if TZ appears and its value is not
a null string, its value is interpreted using rules specific to the
System V and non-System V environments.
.PP
Programs that always wish to use local wall clock time should explicitly
remove the environmental variable TZ with 
.PN unsetenv(3).
.PP
The
.PN ctime
call converts a long integer, pointed to by
.I clock,
representing the time in seconds since
00:00:00 GMT, January 1, 1970,
and returns a pointer to a
26-character string
in the following form.
All the fields have constant width.
.EX
Sun Sep 16 01:03:52 1985\\n\\0
.EE
.PP
The
.PN localtime
and
.PN gmtime
calls return pointers to 
.I tm 
structures, described below.
The
.PN localtime
call corrects for the time zone and possible DST;
.PN gmtime
converts directly to GMT, which is the
time the ULTRIX system uses.
.PP
The
.PN asctime
call converts a 
.I tm 
structure to a 26-character string,
as shown in the previous example,
and returns a pointer
to the string.
.PP
Declarations of all the functions and externals,
and the 
.I tm
structure,
are in the <time.h> header file.
The structure declaration is:
.EX
struct tm {
        int tm_sec;     /\(** seconds (0 - 59) \(**/
        int tm_min;     /\(** minutes (0 - 59) \(**/
        int tm_hour;    /\(** hours (0 - 23) \(**/
        int tm_mday;    /\(** day of month (1 - 31) \(**/
        int tm_mon;     /\(** month of year (0 - 11) \(**/
        int tm_year;    /\(** year \- 1900 \(**/
        int tm_wday;    /\(** day of week (Sunday = 0) \(**/
        int tm_yday;    /\(** day of year (0 - 365) \(**/
        int tm_isdst;   /\(** flag: daylight savings time in
                              effect \(**/
        long tm_gmtoff; /\(** offset from GMT in seconds \(**/
        char *tm_zone;  /\(** abbreviation of timezone name \(**/

};
.EE
.B tm_isdst
is nonzero if DST is in effect.
.PP
\fBtm_gmtoff\fP is the offset (in seconds) of the time represented
from GMT, with positive values indicating East of Greenwich.
.PP
The 
.PN difftime 
call computes the difference between two calendar times:
\fItime1\fP - \fItime0\fP and returns the difference expressed in
seconds.
.PP
The
.PN mktime 
call converts the broken-down local time in the \fItm\fP structure
pointed to by \fItimeptr\fP into a calendar time value with the
same encoding as that of the values returned by 
.PN time .
The values of \fBtm_wday\fP and \fBtm_yday\fP in the structure are
ignored, and the other values are not restricted to the ranges 
indicated above for the \fItm\fP structure. A positive or zero 
value for \fBtm_isdst\fP causes 
.PN mktime
to presume that DST, respectively,
is or is not in effect for the specified time. A negative value causes
.PN mktime 
to attempt to determine whether DST is
in effect for the specified time. On successful completion, 
the values of \fBtm_wday\fP and \fBtm_yday\fP are set appropriately, 
and the other components are set to represent the specified 
calendar time, but with their values forced to the ranges indicated 
above. If the calendar time cannot be represented, the function 
returns the value \fB(time_t)\-1\fP.
.PP
The external variable \fItzname\fP, contains the current time zone
names. The function 
.PN tzset 
sets this variable.
.SS BSD and POSIX Environment Only
If TZ appears in the environment and its value is not a null string,
its value has one of three formats:
.PP
    :
.PP
or
.PP
    :\fIpathname\fP
.PP
or
.PP
    \fIstdoffset[dst[offset][\fP,\fIstart[\fP/\fItime]\fP,\fIend[\fP/\fItime]]]\fP
.PP
If TZ is the single colon format (first format), GMT is used.
.PP
If TZ is the colon followed by a pathname format (second),
the characters following the colon specify a pathname of a
.MS tzfile 5 
format file from which to read the time conversion
information. If the pathname begins with a slash, it represents an
absolute pathname; otherwise the pathname is relative to the system
time conversion information directory 
.PN /etc/zoneinfo. 
If this file fails
for any reason, the GMT offset as provided by the kernel is used.
.PP
If the first character in TZ is not a colon (third format),
the components of the string have the following meaning:
.TP 15
\fIstd\fP and \fIdst\fP
Three or more characters that are the designation for the standard (\fIstd\fP)
or summer (\fIdst\fP) time zone. Only \fIstd\fP is required; if \fIdst\fP
is missing, then summer time does not apply in this locale. Upper- and
lowercase letters are explicitly allowed. Any characters except a leading
colon (:), digits, comma (,), minus (\-), plus (+), and ASCII NUL are
allowed.
.TP 15
\fIoffset\fP
Indicates the value to be added to the local time to arrive at Coordinated
Universal Time. The \fIoffset\fP has the form:
.br
.sp
    \fIhh[:mm[:ss]]\fP
.br
.sp
The minutes (\fImm\fP) and seconds (\fIss\fP) are optional. The hour
(\fIhh\fP) is required and may be a single digit. The \fIoffset\fP
following \fIstd\fP is required. If no \fIoffset\fP follows \fIdst\fP,
summer time is assumed to be one hour ahead of standard time. One or more
digits may be used; the value is always interpreted as a decimal number.
The hour must be between zero and 24, and the minutes (and seconds) \- if
present \- between zero and 59. If preceded by a "\-", the time zone is
east of the Prime Meridian; otherwise it is west (which may be indicated
by an optional preceding "+").
.TP 15
\fIstart\fP and \fIend\fP
Indicates when to change to and back from summer time. \fIStart\fP describes
the date when the change from standard to summer time occurs and \fIend\fP
describes the date when the change back happens. The format of \fIstart\fP
and \fIend\fP must be one of the following:
.RS
.TP 8
J\fIn\fP
The Julian day \fIn\fP (1 \(<= \fIn\fP \(<= 365). Leap days are not counted.
That is, in all years, including leap years, February 28 is day 59 and March
1 is day 60. It is impossible to explicitly refer to the occasional
February 29.
.TP 8
\fIn\fP
The zero-based Julian day (0 \(<= \fIn\fP \(<= 365). Leap days are counted,
and it is possible to refer to February 29.
.TP 8
M\fIm\fP.\fIn\fP.\fId\fP
The \fIn\fPth \fId\fP day of month \fIm\fP (1 \(<= \fIn\fP \(<= 5,
0 \(<= \fId\fP \(<= 6, 1 \(<= \fIm\fP \(<= 12). When \fIn\fP is 5 it refers
to the last \fId\fP day of month \fIm\fP. Day 0 is Sunday.
.RE
.TP 15
\fItime\fP
The \fItime\fP field describes the time when, in current time, the change to
or from summer time occurs. \fITime\fP has the same format as \fIoffset\fP
except that no leading sign (a minus sign (\-) or a plus sign (+))
is allowed. The default, if
\fItime\fP is not given, is 02:00:00.
.PP
As an example of the previous format, if the TZ environment
variable had the value EST5EDT4,M4.1.0,M10.5.0 it would describe the
rule, which went into effect in 1987, for the Eastern time zone in the
USA. Specifically, EST would be the designation for standard time,
which is 5 hours behind GMT. EDT would be the designation for
DST, which is 4 hours behind GMT. DST starts on
the first Sunday in April and ends on the last Sunday in October.
In both cases, since the time was not specified, the change to and from DST would occur at the default time of 2:00 AM.
.PP
The 
.PN timezone
call remains for compatibility reasons only; it is impossible to
reliably map timezone's arguments (\fIzone\fP, a `minutes west of GMT'
value and
\fIdst\fP, a `daylight saving time in effect' flag) to a time zone
abbreviation.
.PP
If the environmental string TZNAME exists, 
.PN timezone 
returns
its value, unless it consists of two comma separated strings, in which
case the second string is returned if \fIdst\fP is non-zero, else
the first string.  If TZNAME does not exist, \fIzone\fP is checked
for equality with a built-in table of values, in which case 
.PN timezone
returns the time zone or daylight time zone abbreviation 
associated with that value.  If the requested \fIzone\fP does 
not appear in the table, the
difference from GMT is returned; that is, in Afghanistan,
.PN timezone(\-(60*4+30), 0)
is appropriate because it is 4:30 ahead of
GMT, and the string `GMT+4:30' is returned.  Programs that in the
past used the 
.PN timezone 
function should return the 
.I zone 
name as
set by 
.PN localtime
to assure correctness.
.SS System V Environment Only
.NXR "ctime subroutine (standard C)" "System V and"
If TZ appears in the environment its value specifies a pathname
of a 
.MS tzfile 5
format file from which to read the time conversion
information. If the pathname begins with a slash, 
it represents an absolute
pathname; otherwise the pathname is relative 
to the system time conversion
information directory 
.PN /etc/zoneinfo .
.PP
If TZ appears in the environment and using the value as a pathname
of a 
.MS tzfile 5
format file fails for any reason, the value is assumed
to be a three-letter time zone name followed by a number 
representing the difference between local 
time and GMT in hours, followed
by an optional three-letter name for a time zone 
on DST.
For example, the setting for New Jersey would be EST5EDT.
.SS System V and POSIX Environment Only
The external \fIlong\fP variable 
.PN timezone
contains the difference,
in seconds, between GMT and local standard time (in EST, 
.PN timezone
is 5*60*60).
The external variable \fIdaylight\fP is nonzero if any daylight time
rules are to be applied (that is, if the specified timezone
includes rules for daylight time).
These
variables are set whenever
.PN tzset ,
.PN ctime ,
.PN localtime ,
.PN mktime ,
or
.PN strftime
are called.
.SH Restrictions
The return values point to static data whose content is overwritten by
each call.  The \fBtm_zone\fP field of a returned \fBstruct tm\fP
points to a static array of characters, which will also be overwritten
at the next call (and by calls to 
.PN tzset).
.SH Files
.ta \w'/etc/zoneinfo/localtime\0\0'u
/etc/zoneinfo	time zone information directory
.br
/etc/zoneinfo/localtime	local time zone file
.SH See Also
gettimeofday(2), getenv(3), strftime(3), time(3), tzfile(5), environ(7)


lib/libc/gen/ctime.c
/*
 * Copyright (c) 1987 Regents of the University of California.
 * All rights reserved.  The Berkeley software License Agreement
 * specifies the terms and conditions for redistribution.
 */

#ifndef lint
static	char	*sccsid = "@(#)ctime.c	4.1	(ULTRIX)	7/3/90";
#endif lint

/************************************************************************
 *									*
 *			Copyright (c) 1989 by				*
 *		Digital Equipment Corporation, Maynard, MA		*
 *			All rights reserved.				*
 *									*
 *   This software is furnished under a license and may be used and	*
 *   copied  only  in accordance with the terms of such license and	*
 *   with the  inclusion  of  the  above  copyright  notice.   This	*
 *   software  or  any  other copies thereof may not be provided or	*
 *   otherwise made available to any other person.  No title to and	*
 *   ownership of the software is hereby transferred.			*
 *									*
 *   This software is  derived  from  software  received  from  the	*
 *   University    of   California,   Berkeley,   and   from   Bell	*
 *   Laboratories.  Use, duplication, or disclosure is  subject  to	*
 *   restrictions  under  license  agreements  with  University  of	*
 *   California and with AT&T.						*
 *									*
 *   The information in this software is subject to change  without	*
 *   notice  and should not be construed as a commitment by Digital	*
 *   Equipment Corporation.						*
 *									*
 *   Digital assumes no responsibility for the use  or  reliability	*
 *   of its software on equipment which is not supplied by Digital.	*
 *									*
 ************************************************************************/
/************************************************************************
 *			Modification History
 *
 * 004	Jon Reeves, 20-Nov-1989
 *	Added _last_tz_set logic: effectively do tzset every time, not
 *	just first, since environment may have changed.  Still leaves a
 *	little acceleration in place.  Believe user's tm_isdst for
 *	mktime.
 *
 * 003	Ken Lesniak, 31-Aug-1989
 *	Modified mktime to correctly handle all times not representable
 *
 * 002	Ken Lesniak, 17-Jul-1989
 *	Moved tz_is_set flag to tzset.c and actually set flag in tzset()
 *
 * 001	Ken Lesniak, 20-Mar-1989
 *	Added support for POSIX 1003.1 compliancy.
 *	Added mktime and difftime routines for ANSI X3J11.
 *	Moved tzset* routines to tzset.c.
 *
 *	Based on Berkeley ctime.c 1.2
 ************************************************************************/

#include <sys/time.h>
#include <tzfile.h>
#include <limits.h>
#include <stdlib.h>
#include <string.h>
#include "tzs.h"

char *
ctime(t)
time_t *t;
{
	return(asctime(localtime(t)));
}

static char *
itoa2(cp, n)
register char *cp;
int	n;
{
	if (n >= 10)
		*cp++ = (n / 10) % 10 + '0';
	else
		*cp++ = ' ';
	*cp++ = n % 10 + '0';
	return cp;
}

char *
asctime(t)
register struct tm *	t;
{
	static char	wday_name[DAYS_PER_WEEK][3] = {
		"Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"
	};
	static char	mon_name[MONS_PER_YEAR][3] = {
		"Jan", "Feb", "Mar", "Apr", "May", "Jun",
		"Jul", "Aug", "Sep", "Oct", "Nov", "Dec"
	};
	static char	result[26];
	register char	*cp;
	register char	*ncp;
	register int	year;

	cp = result;
	ncp = wday_name[t->tm_wday];
	*cp++ = *ncp++;
	*cp++ = *ncp++;
	*cp++ = *ncp++;

	*cp++ = ' ';
	ncp = mon_name[t->tm_mon];
	*cp++ = *ncp++;
	*cp++ = *ncp++;
	*cp++ = *ncp++;

	*cp++ = ' ';
	cp = itoa2(cp, t->tm_mday);

	*cp++ = ' ';
	cp = itoa2(cp, t->tm_hour + 100);

	*cp++ = ':';
	cp = itoa2(cp, t->tm_min + 100);

	*cp++ = ':';
	cp = itoa2(cp, t->tm_sec + 100);

	*cp++ = ' ';
	year = TM_YEAR_BASE + t->tm_year;
	cp = itoa2(cp, year / 100);
	cp = itoa2(cp, year);

	*cp++ = '\n';
	*cp++ = '\0';

	return result;
}

struct tm *		offtime();

extern struct state	_tzs;

extern int		_tz_is_set;
extern char 		_last_tz_set[LAST_TZ_LEN];

struct tm *
localtime(timep)
time_t *	timep;
{
	register struct ttinfo *	ttisp;
	register struct tm *		tmp;
	register int			i;
	time_t				t;
	char *				tz;

	if (!_tz_is_set || (tz=getenv("TZ")) && strcmp(tz, _last_tz_set))
		(void) tzset();

	t = *timep;
	if (_tzs.timecnt == 0 || t < _tzs.ats[0]) {
		i = 0;
		while (_tzs.ttis[i].tt_isdst)
			if (++i >= _tzs.timecnt) {
				i = 0;
				break;
			}
	} else {
		for (i = 1; i < _tzs.timecnt; ++i)
			if (t < _tzs.ats[i])
				break;
		i = _tzs.types[i - 1];
	}
	ttisp = &_tzs.ttis[i];
	/*
	** To get (wrong) behavior that's compatible with System V Release 2.0
	** you'd replace the statement below with
	**	tmp = offtime((time_t) (t + ttisp->tt_gmtoff), 0L);
	*/
	tmp = offtime(&t, ttisp->tt_gmtoff);
	tmp->tm_isdst = ttisp->tt_isdst;
	tzname[tmp->tm_isdst] = &_tzs.chars[ttisp->tt_abbrind];
	tmp->tm_zone = &_tzs.chars[ttisp->tt_abbrind];
	return tmp;
}

struct tm *
gmtime(clock)
time_t *	clock;
{
	register struct tm *	tmp;

	tmp = offtime(clock, 0L);
	tzname[0] = "GMT";
	tmp->tm_zone = "GMT";		/* UCT ? */
	return tmp;
}

static int	mon_lengths[2][MONS_PER_YEAR] = {
	31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31,
	31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31
};

static int	year_lengths[2] = {
	DAYS_PER_NYEAR, DAYS_PER_LYEAR
};

struct tm *
offtime(clock, offset)
time_t *	clock;
long		offset;
{
	register struct tm *	tmp;
	register long		days;
	register long		rem;
	register int		y;
	register int		yleap;
	register int *		ip;
	static struct tm	tm;

	tmp = &tm;
	days = *clock / SECS_PER_DAY;
	rem = *clock % SECS_PER_DAY;
	rem += offset;
	while (rem < 0) {
		rem += SECS_PER_DAY;
		--days;
	}
	while (rem >= SECS_PER_DAY) {
		rem -= SECS_PER_DAY;
		++days;
	}
	tmp->tm_hour = (int) (rem / SECS_PER_HOUR);
	rem = rem % SECS_PER_HOUR;
	tmp->tm_min = (int) (rem / SECS_PER_MIN);
	tmp->tm_sec = (int) (rem % SECS_PER_MIN);
	tmp->tm_wday = (int) ((EPOCH_WDAY + days) % DAYS_PER_WEEK);
	if (tmp->tm_wday < 0)
		tmp->tm_wday += DAYS_PER_WEEK;
	y = EPOCH_YEAR;
	if (days >= 0)
		for ( ; ; ) {
			yleap = isleap(y);
			if (days < (long) year_lengths[yleap])
				break;
			++y;
			days -= (long) year_lengths[yleap];
		}
	else do {
		--y;
		yleap = isleap(y);
		days += (long) year_lengths[yleap];
	} while (days < 0);
	tmp->tm_year = y - TM_YEAR_BASE;
	tmp->tm_yday = (int) days;
	ip = mon_lengths[yleap];
	for (tmp->tm_mon = 0; days >= (long) ip[tmp->tm_mon]; ++(tmp->tm_mon))
		days -= (long) ip[tmp->tm_mon];
	tmp->tm_mday = (int) (days + 1);
	tmp->tm_isdst = 0;
	tmp->tm_zone = "";
	tmp->tm_gmtoff = offset;
	return tmp;
}


double difftime(time1, time0)
	time_t		time1;
	time_t		time0;
    {
	return (double)time1 - (double)time0;
    }


/*
 * The following types and routines are provided so that mktime() can
 * use quadword math as it combines the broken-down time into
 * a time_t value. mktime() does not impose any bounds on the broken-down
 * time components, without the quadword math, all sorts of overflows
 * are possible.
 *
 * The routines are not general purpose quadword routines. The types
 * of the operands are restricted, to allow shortcuts in the algorthims.
 *
 * I know this is ugly, I can't think of any other way to allow mktime
 * to handle all broken-down times that can represent a valid time_t.
 */

typedef union {
	unsigned long	lw[2];
	unsigned short	w[4];
    } quadword;

typedef union {
	unsigned long	lw;
	unsigned short	w[2];
    } longword;

#define ULONG unsigned long

/* quadword(qw) = quadword(qw) + long(op) */

void qadd(qw, op)
	register quadword *qw;
	long		op;
    {
	longword	ac;
	longword	sext;
	longword	lw_op;

	sext.lw = 0;
	if (op < 0)
	    sext.lw = (unsigned long)-1;

	lw_op.lw = (ULONG)op;

	qw->w[0] = (ac.lw = (ULONG)qw->w[0] + lw_op.w[0]);
	qw->w[1] = (ac.lw = (ULONG)qw->w[1] + lw_op.w[1] + (ULONG)ac.w[1]);
	qw->w[2] = (ac.lw = (ULONG)qw->w[2] + (ULONG)sext.w[0] + (ULONG)ac.w[1]);
	qw->w[3] += (ULONG)sext.w[1] + (ULONG)ac.w[1];
    };

/* quadword(qw) = (quadword(qw) * unsigned short(mop)) + long(aop) */

void qmuladd(qw, mop, aop)
	register quadword *qw;
	unsigned short	mop;
	long		aop;
    {
	longword	ac;

	qw->w[0] = (ac.lw = (ULONG)qw->w[0] * (ULONG)mop);
	qw->w[1] = (ac.lw = ((ULONG)qw->w[1] * (ULONG)mop) + ac.w[1]);
	qw->w[2] = (ac.lw = ((ULONG)qw->w[2] * (ULONG)mop) + ac.w[1]);
	qw->w[3] = (qw->w[3] * mop) + ac.w[1];

	qadd(qw, aop);
    }

time_t mktime(tp)
	register struct tm *tp;
    {
	time_t		t;
	quadword	tq;
	int		y;
	int		m;
	int		ly;
	int		isdst;
	int		i;
	struct ttinfo	*ttis;
	char		*tz;

#define mktime_max_year (INT_MAX + (INT_MIN / MONS_PER_YEAR) + \
	(INT_MIN / DAYS_PER_NYEAR) + \
	(INT_MIN / (HOURS_PER_DAY * DAYS_PER_NYEAR)) + \
	(INT_MIN / (MINS_PER_HOUR * HOURS_PER_DAY * DAYS_PER_NYEAR)) + \
	(INT_MIN / (SECS_PER_MIN * MINS_PER_HOUR * HOURS_PER_DAY * DAYS_PER_NYEAR)) + \
	TM_YEAR_BASE)
#define mktime_min_year (INT_MIN + (INT_MAX / MONS_PER_YEAR) + \
	(INT_MAX / DAYS_PER_NYEAR) + \
	(INT_MAX / (HOURS_PER_DAY * DAYS_PER_NYEAR)) + \
	(INT_MAX / (MINS_PER_HOUR * HOURS_PER_DAY * DAYS_PER_NYEAR)) + \
	(INT_MAX / (SECS_PER_MIN * MINS_PER_HOUR * HOURS_PER_DAY * DAYS_PER_NYEAR)) + \
	TM_YEAR_BASE)

	/* Make sure the year is in bounds to avoid overflowing the */
	/* month-year normalization */

	if (tp->tm_year < mktime_min_year || tp->tm_year > mktime_max_year)
	    return (time_t)-1;

	/* Get normalized month and year */

	y = tp->tm_year + TM_YEAR_BASE + (tp->tm_mon / MONS_PER_YEAR);
	if ((m = tp->tm_mon % MONS_PER_YEAR) < 0) {
	    m += MONS_PER_YEAR;
	    y--;
	};
	ly = isleap(y);

	/* Determine the number of days since the epoch */

	tq.lw[0] = tq.lw[1] = -1;
	qadd(&tq, tp->tm_mday);
	while (m != 0)
	    qadd(&tq, mon_lengths[ly][--m]);

	if (y >= EPOCH_YEAR) {
	    while (y != EPOCH_YEAR) {
		y--;
		qadd(&tq, year_lengths[isleap(y)]);
	    };
	} else {
	    do {
		qadd(&tq, -year_lengths[isleap(y)]);
		y++;
	    } while (y != EPOCH_YEAR);
	};
	
	/* Now convert everything to seconds */

	qmuladd(&tq, HOURS_PER_DAY, tp->tm_hour);
	qmuladd(&tq, MINS_PER_HOUR, tp->tm_min);
	qmuladd(&tq, SECS_PER_MIN, tp->tm_sec);

	/* Make sure the resulting time can be represented by a time_t */

	if (tq.lw[1] != 0 && tq.lw[1] != (unsigned long)-1)
	    return (time_t)-1;

	t = (time_t)tq.lw[0];

	/* Determine the GMT offset for the time */

	if (!_tz_is_set || (tz=getenv("TZ")) && strcmp(tz, _last_tz_set))
	    (void)tzset();

	ttis = &_tzs.ttis[0];
	if (tp->tm_isdst < 0) {
	    if (_tzs.timecnt != 0 && t >= _tzs.ats[0] + ttis->tt_gmtoff) {
		for (i = 1; i < _tzs.timecnt; i++) {
		    ttis = &_tzs.ttis[_tzs.types[i - 1]];
		    if (t < _tzs.ats[i - 1] + ttis->tt_gmtoff)
			return (time_t)-1;
		    if (t < _tzs.ats[i] + ttis->tt_gmtoff)
			break;
		};
	    };
	} else {
	    /* User specified DST, normalize to 0 or 1 (yes, '=' is right) */
	    if (isdst = tp->tm_isdst)
		isdst = 1;

	    /* Search for an appropriate DST setting; fall out if no DST */
	    for (i = 0; i < _tzs.timecnt; i++) {
		ttis = &_tzs.ttis[_tzs.types[i]];
		if (ttis->tt_isdst == isdst)
		    break;
	    }

	    /* Asked for DST (or standard), but none defined */
	    if (ttis->tt_isdst != isdst)
		return (time_t)-1;
	}

	/* Adjust for time zone and daylight savings time */

	t -= ttis->tt_gmtoff;

	/* Normalize the broken down time and fill in the unspecified fields */

	(void)memcpy(tp, localtime(&t), sizeof(struct tm));

	/* Return the time */

	return t;
    }
